core/src/main/java/io/grpc/Metadata.java

```diff
@@ -132,87 +135,145 @@ public Integer parseAsciiString(String serialized) {
     }
   };
 
-  /** All value lists can be added to. No value list may be empty. */
-  // Use LinkedHashMap for consistent ordering for tests.
-  private final Map<String, List<MetadataEntry>> store =
-      new LinkedHashMap<String, List<MetadataEntry>>();
-
-  /** The number of headers stored by this metadata.  */
-  private int storeCount;
+  /**
+   * Constructor called by the transport layer when it receives binary metadata.
+   */
+  Metadata(byte[]... binaryValues) {
+    this(binaryValues.length / 2, binaryValues);
+  }
 
   /**
    * Constructor called by the transport layer when it receives binary metadata.
    */
   // TODO(louiscryan): Convert to use ByteString so we can cache transformations
-  @Internal
-  public Metadata(byte[]... binaryValues) {
-    checkArgument(binaryValues.length % 2 == 0,
-        "Odd number of key-value pairs: %s", binaryValues.length);
-    for (int i = 0; i < binaryValues.length; i += 2) {
-      // The transport might provide an array with null values at the end.
-      if (binaryValues[i] == null) {
-        break;
-      }
-      String name = new String(binaryValues[i], US_ASCII);
-      storeAdd(name, new MetadataEntry(name.endsWith(BINARY_HEADER_SUFFIX), binaryValues[i + 1]));
-    }
+  Metadata(int usedNames, byte[]... binaryValues) {
+    // TODO(carl-mastrangelo): maybe remove this, unless
+    // https://bugs.openjdk.java.net/browse/JDK-6445664 is fixed.
+    assert (binaryValues.length & 1) == 0 : "Odd number of key-value pairs " + binaryValues.length;
+    size = usedNames;
+    namesAndValues = binaryValues;
+    for (int i = 0; i < usedNames * 2; i++) {
+      checkNotNull(namesAndValues[i], "val was null %s/%s", i, namesAndValues.length);
+    }
+    typedValues = size != 0 ? new Object[size] : null;
```

Comment:
I suppose it could, but I would be surprised.  This method is invoked on receipt of tranport headers, which almost always contains /some/ headers.  


I suppose I could always allocate and shrink this method.